core and prov/shm: Add HMEM IPC Caching Mechanism.

shijin-aws opened the pull request:
~~This PR is based on the commit https://github.com/ofiwg/libfabric/pull/7848/commits/d34eacd16de58e45bfb256d5ae974c1c22cb8953 in the draft PR https://github.com/ofiwg/libfabric/pull/7848 opened by @amirshehataornl~~


Opening and closing the HMEM IPC handles are expensive operations. They
    should not be performed on every data operation.
    
This patch adds a caching mechanism. On the very first IPC handle open
    call, the address is cached and the mapped memory address is stored. In
    subsequent calls the cache is queried and the mapped address is retrieved.
    
The IPC cache is implemented as an object of ofi_mr_cache, it has the following
    cache attributes:
    
 cache->add_region: open the ipc handle
 cache->delete_region: close the ipc handle
 cache->monitor: a new ipc monitor of type struct ofi_mem_monitor to check whether the
    ipc_handle is stale.

tschuett commented on include/pgtable.h:
```diff
@@ -0,0 +1,256 @@
+/**
```
Disclaimer: I am not a lawyer. There is a similar file in the UCX repository: ucx/src/ucs/datastruct/pgtable.h.
Do you believe that licenses are compatible?

tschuett commented on include/pgtable.h:
```diff
@@ -0,0 +1,256 @@
+/*
+ * Assertions which are checked in compile-time
+ *
+ * Usage: UCS_STATIC_ASSERT(condition)
```
UCS_STATIC_ASSERT is an UCX term.

amirshehataornl replied:
I think I mentioned this somewhere I adopted the caching mechanism from the UCX caching mechanism. As far as I understand the licences it does allow usage as long as the attribution is kept. But I'm not a lawyer either. Probably should remove line 19 though and clean up any other similar things.

here is their license file:

Copyright (c) 2014-2015      UT-Battelle, LLC. All rights reserved.
Copyright (C) 2014-2020      Mellanox Technologies Ltd. All rights reserved.
Copyright (C) 2014-2015      The University of Houston System. All rights reserved.
Copyright (C) 2015           The University of Tennessee and The University 
                             of Tennessee Research Foundation. All rights reserved.
Copyright (C) 2016-2020      ARM Ltd. All rights reserved.
Copyright (c) 2016           Los Alamos National Security, LLC. All rights reserved.
Copyright (C) 2016-2020      Advanced Micro Devices, Inc.  All rights reserved.
Copyright (C) 2019           UChicago Argonne, LLC.  All rights reserved.
Copyright (c) 2018-2020      NVIDIA CORPORATION. All rights reserved.
Copyright (C) 2020           Huawei Technologies Co., Ltd. All rights reserved.
Copyright (C) 2016-2020      Stony Brook University. All rights reserved.

Redistribution and use in source and binary forms, with or without 
modification, are permitted provided that the following conditions 
are met:

1. Redistributions of source code must retain the above copyright 
notice, this list of conditions and the following disclaimer.
2. Redistributions in binary form must reproduce the above copyright
notice, this list of conditions and the following disclaimer in the 
documentation and/or other materials provided with the distribution.
3. Neither the name of the copyright holder nor the names of its 
contributors may be used to endorse or promote products derived from 
this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR 
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT 
HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, 
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED 
TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR 
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF 
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING 
NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS 
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

shijin-aws replied:
Thank you! Will do.

wckzhang commented on fabtests/common/hmem_cuda.c:
```diff
@@ -49,15 +50,55 @@ struct cuda_ops {
+{
+	int true_flag = 1;
+	CUresult cu_result;
+	cu_result = cuda_ops.cuPointerSetAttribute((void *) &true_flag,
```
Didn't you say this was insufficient anyway? So device synchronization is being done instead?

shijin-aws replied:
This is still required to ensure the synchronous data movement between devices. Because we only do synchronize in libfabric if the buffer is not allocated/registered. For cudaMemcpy between device buffers, there is no cudaDeviceSynchronize inside libfabric.

wckzhang commented on prov/shm/src/smr_progress.c:
```diff
@@ -513,11 +511,11 @@ static int smr_progress_ipc(struct smr_cmd *cmd, enum fi_hmem_iface iface,
-
-	/* Truncation error takes precedence over close_handle error */
-	ret = ofi_hmem_close_handle(cmd->msg.data.ipc_info.iface, base);
+		/* Truncation error takes precedence over close_handle error */
```
Was this supposed to be encapsulated within the if block? Is this a bugfix or a bug?

shijin-aws replied:
So before this PR, the IPC handle is closed every time after data copy (for both ZE and non-ZE). But with this PR, we do not want to close handle each time. ZE is an exception because this PR does not want to modify ZE's behavior, so the close is moved to the if block.

j-xiong replied:
@aingerson ZE behaves differently because of the use of ze_hmem_open_shared_handle(). Can ZE be adapted to take advantage of the IPC cache? for example, moving the call of ze_hmem_open_shared_handle() into the IPC cache code, or using the shared handle as the IPC handle? Can be done as a follow up PR.

aingerson replied:
Sorry for some reason I didn't see this earlier. We can take advantage of the cache to a certain extent. The handles are fds so we need to make sure to take into consideration having too many fds open. But I think this can be done. I can follow up after this is merged.

shijin-aws commented:
continuous-integration/appveyor/pr  breaks due to `bitops.h` cannot be compiled on Windows, will address in a newer revision

```
"C:\projects\libfabric\libfabric.sln" (default target) (1) ->
"C:\projects\libfabric\libfabric.vcxproj" (default target) (2) ->
(ClCompile target) -> 
  c:\projects\libfabric\include\bitops.h(21): fatal error C1189: #error:  "Unsupported architecture" [C:\projects\libfabric\libfabric.vcxproj]
  c:\projects\libfabric\include\bitops.h(21): fatal error C1189: #error:  "Unsupported architecture" [C:\projects\libfabric\libfabric.vcxproj]
    466 Warning(s)
    2 Error(s)

```

j-xiong commented on include/hmem_cache.h:
```diff
@@ -0,0 +1,82 @@
+	pgt_region_t		super;
+	struct dlist_entry	list;
+	struct ipc_info		key;
+	void				*mapped_addr;
```
alignment is off.

j-xiong commented on include/hmem_cache.h:
```diff
@@ -0,0 +1,82 @@
+};
+
+int ipc_create_hmem_cache(struct hmem_cache **cache,
+						  const char *name);
```
please align 'const' with 'struct'.

j-xiong commented on include/hmem_cache.h:
```diff
@@ -0,0 +1,82 @@
+						  const char *name);
+void ipc_destroy_hmem_cache(struct hmem_cache *cache);
+int ipc_cache_map_memhandle(struct hmem_cache *cache, struct ipc_info *key,
+							void **mapped_addr);
```
check alignment.

j-xiong commented on include/ofi.h:
```diff
@@ -149,8 +149,21 @@ extern "C" {
 		_a > _b ? _a : _b; })
 #endif
 
+#define BIT(i) (1ul << (i))
```
BIT is already defined 20~ lines later in the same file.

shijin-aws replied:
Fixed.

shijin-aws replied:
Edited. Commented by mistake

j-xiong commented on src/pgtable.c:
```diff
@@ -0,0 +1,602 @@
+ */
+static int
+pgtable_insert_page(pgtable_t *pgtable, pgt_addr_t address,
+						unsigned order, pgt_region_t *region)
```
check alignment.

j-xiong commented on src/pgtable.c:
```diff
@@ -0,0 +1,602 @@
+ */
+static int
+pgtable_remove_page_recurs(pgtable_t *pgtable, pgt_addr_t address,
+							   unsigned order, pgt_dir_t *pgd,
+							   pgt_entry_t *pte, unsigned shift,
+							   pgt_region_t *region)
```
check alignment.

j-xiong commented on src/pgtable.c:
```diff
@@ -0,0 +1,602 @@
+		next_pte   = &next_dir->entries[(address >> next_shift) & PGT_ENTRY_MASK];
+
+		status = pgtable_remove_page_recurs(pgtable, address, order, next_dir,
+												next_pte, next_shift, region);
```
alignment if way off.

j-xiong commented on src/pgtable.c:
```diff
@@ -0,0 +1,602 @@
+
+static int
+pgtable_remove_page(pgtable_t *pgtable, pgt_addr_t address,
+						unsigned order, pgt_region_t *region)
```
alignment.

j-xiong commented on src/pgtable.c:
```diff
@@ -0,0 +1,602 @@
+	}
+
+	status = pgtable_remove_page_recurs(pgtable, address, order, &dummy_pgd,
+											&pgtable->root, pgtable->shift,
+											region);
```
alignment.

j-xiong commented on src/pgtable.c:
```diff
@@ -0,0 +1,602 @@
+}
+
+pgt_region_t *pgtable_lookup(const pgtable_t *pgtable,
+									 pgt_addr_t address)
```
alignment.

j-xiong commented on src/pgtable.c:
```diff
@@ -0,0 +1,602 @@
+}
+
+static void pgtable_search_recurs(const pgtable_t *pgtable,
+									  pgt_addr_t address, unsigned order,
+									  const pgt_entry_t *pte, unsigned shift,
+									  pgt_search_callback_t cb, void *arg,
+									  pgt_region_t **last_p)
```
alignment.

j-xiong commented on src/pgtable.c:
```diff
@@ -0,0 +1,602 @@
+			assert(order <= next_shift);
+			next_pte = &dir->entries[(address >> next_shift) & PGT_ENTRY_MASK];
+			pgtable_search_recurs(pgtable, address, order, next_pte,
+									  next_shift, cb, arg, last_p);
```
alignment.

j-xiong commented on src/pgtable.c:
```diff
@@ -0,0 +1,602 @@
+			for (i = 0; i < PGT_ENTRIES_PER_DIR; ++i) {
+				next_pte = &dir->entries[i];
+				pgtable_search_recurs(pgtable, address, order, next_pte,
+										  next_shift, cb, arg, last_p);
```
alignment.

j-xiong commented on src/pgtable.c:
```diff
@@ -0,0 +1,602 @@
+}
+
+void pgtable_search_range(const pgtable_t *pgtable,
+							  pgt_addr_t from, pgt_addr_t to,
+							  pgt_search_callback_t cb, void *arg)
```
alignment.

j-xiong commented on src/pgtable.c:
```diff
@@ -0,0 +1,602 @@
+		order = pgtable_get_next_page_order(address, end);
+		if ((address & pgtable->mask) == pgtable->base) {
+			pgtable_search_recurs(pgtable, address, order, &pgtable->root,
+									  pgtable->shift, cb, arg, &last);
```
alignment.

j-xiong commented on src/pgtable.c:
```diff
@@ -0,0 +1,602 @@
+}
+
+static void pgtable_purge_callback(const pgtable_t *pgtable,
+									   pgt_region_t *region,
+									   void *arg)
```
alignment.

j-xiong commented on src/pgtable.c:
```diff
@@ -0,0 +1,602 @@
+}
+
+void pgtable_purge(pgtable_t *pgtable, pgt_search_callback_t cb,
+					   void *arg)
```
alignment.

j-xiong commented on src/pgtable.c:
```diff
@@ -0,0 +1,602 @@
+	from = pgtable->base;
+	to   = pgtable->base + ((1ul << pgtable->shift) & pgtable->mask) - 1;
+	pgtable_search_range(pgtable, from, to, pgtable_purge_callback,
+							 &next_region);
```
alignment.

j-xiong commented on src/pgtable.c:
```diff
@@ -0,0 +1,602 @@
+}
+
+int pgtable_init(pgtable_t *pgtable,
+				pgt_dir_alloc_callback_t alloc_cb,
+				pgt_dir_release_callback_t release_cb)
```
alignment.

j-xiong commented on src/pgtable.c:
```diff
@@ -0,0 +1,602 @@
+	/* If there is just one PTE, we can reduce the page table to map
+	 * this PTE only.
+	 */
+	if (pgd->count != 1) {
```
No need for braces over a single statement, unless for matching `if ... else ...` structure. There are many similar occurrences in the file.

wckzhang replied:
@j-xiong not the author but I personally prefer braces over single line if statements anyway. Leads to less accidents when adding new lines to the if statements (adding a new line w/o braces will lead to some hard to find bugs but it will still compile)

j-xiong replied:
@wckzhang Although that is personal preference, the coding convention here is to not use braces over a single line. Braces are added when new lines are added -- not too difficult to follow.

No coding convention can eliminate coding accident. Sticking with the same convention does help improve code readability in the long run.

shefty commented on src/hmem_cache.c:
```diff
@@ -0,0 +1,309 @@
+
+	ret = pgtable_init(&cache_desc->pgtable,
+			   ipc_cache_pgt_dir_alloc,
+			   ipc_cache_pgt_dir_release);
```
Why do we need to setup callbacks?

shefty commented on include/pgtable.h:
```diff
@@ -0,0 +1,254 @@
+	unsigned                   shift;       /**< page table address span is 2**shift */
+	unsigned                   num_regions; /**< total number of regions */
+	pgt_dir_alloc_callback_t   pgd_alloc_cb;
+	pgt_dir_release_callback_t pgd_release_cb;
```
Where are these cb functions used?

shefty commented on src/pgtable.c:
```diff
@@ -0,0 +1,588 @@
+ * Copyright (C) 2022 Amazon.com, Inc. or its affiliates.
+ *
+ * See file LICENSE for terms.
+ */
```
This seems overly complex for what we need, with callbacks where simple function calls should be use instead. We already have code in tree for managing a cache.  Why is that code not used?

shijin-aws replied:
I will re-work the PR to use rbmap which is used when managing mr cache. That will get rid of the current pgtable stuff that is copied from UCX, and also gets rid of the license concern we have.

tschuett replied:
Is the rbmap as efficient as the pgtable or is their room for improvement?
The first commit hints at it being fairly optimized for their use case.
https://github.com/openucx/ucx/commit/9dbd386eb3f08bb0512eb39bb52450bfcf75f0fa

amirshehataornl replied:
I can probably run a comparison if we have an implementation with rbmap. 

When I adopted the pgtable caching my tests showed a good performance boost for SHM testing, because we no longer had to open/close the shared handled on every write/read, which were pretty expensive. So the first time, it would get a cache miss, then on every other iteration of the OSU test it would be a cache hit. However, most of the performance was obtained when I moved to asynchronous memory copies, which I implemented for ROCM. I did an implementation for both HSA and HIP. I'm assuming HIP style asynchronous memcpy can work for CUDA (just an educated guess, never looked at cuda)

The difference in performance between the pgtable and rbmap implementation would be in the handle lookup.  Note, I also used the same mechanism for XPMEM support

shijin-aws replied:
@amirshehataornl I observed 30x time perf boost when switching from no-cache to cache with the current pgtable implementation, for CUDA IPC when running shm test. I can redo such perf test after I implement the cache with rbmap.

amirshehataornl replied:
@shijin-aws thanks for the info.. I'd like to do the same for rocm

shefty replied:
The ofi_mr_cache currently uses an rbtree, but that's hidden from the user.  There's no reason why it couldn't change to some other lookup method.  My question was is there a reason that cannot be used, either as-is or by modifying it?  There's already support for hmem as part of the ofi_mr_entry.

I completely believe that a cache is faster.

shijin-aws replied:
@shefty Yeah you are right, we should be able to reuse the same lookup method for the ipc cache. One question raised earlier is whether this lookup method (rbtree) is more performant than the pgtable in UCX. We can run some comparison for it.

shefty replied:
The rbtree implementation was just easy.  We can switch to a hash table or other lookup method.  It makes sense to me to start with what we have and update the ofi_mr_cache separately to improve the performance.  That way, all cached entries benefit from a faster lookup, rather than only IPC related GPU buffers.

shijin-aws replied:
I am a bit uncertain whether we should use `ofi_mr_cache` for IPC cache. They are different things right? You don't need a memory registration to use IPC? and I don't see whether a memory monitor is really needed for IPC cache. The IPC handle being opened must be valid because it's just got by the peer process.

shefty replied:
We are caching memory regions (identified by an iovec -- address + length), and storing with the region some metadata.  The current metadata includes the iovec and possible device interface and id, plus some provider specific data.  What we use the cache for doesn't really matter.  It's an address + length lookup to return some data structure.  If the memory monitor always returns that everything in the cache is always valid, then that's easy.  It would definitely be surprising for items that are cached, but if it works, that's fine.

shijin-aws replied:
Makes sense, thanks for the clarification!

wckzhang replied:
@shefty If we're going to use the mr cache as a generic iov cache, I wonder if this will necessitate the bringing back of merge regions as an option of the cache

shefty replied:
Maybe, but I doubt it will be necessary.  Merging resulted in worse performance when we actually needed to do something as part of the merge (for example, register a new memory region to replace the ones being merged).

I thought in the case of HMEM, we retrieved the base address of the allocated region using the target address as input.  So, the cached region should be relative to the base address, not the address passed in by the app.

wckzhang replied:
In the design I mocked up to see how it would work, I set the stored base addr and len to be the base address of an allocation and 1 (as that would be how CUDA works), in which case merge regions wouldn't be needed. However, I didn't know if all ipc handles maintain the same behavior of base ptr address. The UCX implementation for example does base and len, and invalidates regions based on overlap iirc. If we can just use [base_addr, 1] as the key, then it shouldn't be too difficult to do.

shefty replied:
I'm guessing there's no way to get the actual size of the allocated region, just the base address?

shijin-aws replied:
for cuda, you can do cuPointerGetAttribute, to get the base address, and the length?

j-xiong replied:
The actual allocation length is also available with oneAPI L0.

shijin-aws commented:
@shefty We are trying to maximally reuse the current `ofi_mr_cache` framework to implement IPC cache. The current design is to make the cache has add/delete_region ops as the ipc_handle_open/close, and create an IPC handle monitor of type `ofi_mem_monitor` to validate an IPC handle when it's retrieved from the mr cache.

According to CUDA documentation

>If a region of memory is freed with [cudaFree](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1ga042655cbbf3408f01061652a075e094) and a subsequent call to [cudaMalloc](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1g37d37965bfb4803b6d4e59ff26856356) returns memory with the same device address, [cudaIpcGetMemHandle](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g8a37f7dfafaca652391d0758b3667539) will return a unique handle for the new memory.

So it is possible that receiver gets a different IPC handle for the same memory address (and length) from the sender. Then the mr entry with the same iov but the old IPC handle should be uncached, and the new mr entry that has the new IPC handle should be inserted to the cache. This is how `ofi_mr_cache_search` currently does so it meets our demand as long as we can implement the monitor correctly.

The current `ofi_mem_monitor` has the following API for the valid (similar for subscribe, etc) ops:

```
bool (*valid)(struct ofi_mem_monitor *notifier, const void *addr,
		      size_t len, union ofi_mr_hmem_info *hmem_info);
```

To realize the comparison between the new IPC handle and the old IPC handle, we are thinking of modifying this API in two possible ways:

1. Adding the `ipc_handle` as a new member in the `ofi_mr_hmem_info`,  and add a new `ipc_handle` argument into the API as

```
bool (*valid)(struct ofi_mem_monitor *notifier, const void *addr,
		      size_t len, void **ipc_handle
                      union ofi_mr_hmem_info *hmem_info);
```

Then comparison is  performed between `ipc_handle` (new handle) and `hmem_info->ipc_handle` (old handle)


2. Modifying the API  into 
```
bool (*valid)(struct ofi_mem_monitor *notifier, struct ofi_mr_info *info, struct ofi_mr_entry *entry);
```
and make `ipc_handle` part of `struct ofi_mr_info`.

Then the comparison is performed between `info->ipc_handle` and `entry->info->ipc_handle`. The `info` object is constructed inside `ofi_mr_cache_search` that contains the iov and ipc_handle of the mr_entry to be searched. The `entry` object is returned by the `ofi_rbt_find` in `ofi_mr_cache_search` that is the found mr entry.


We are concerned that if such monitor API change hurts, considering the current import monitor also follows this API and the `fi_ops_mem_monitor` is an external API

```
struct fi_ops_mem_monitor {
	size_t	size;
	int	(*start)(struct fid_mem_monitor *monitor);
	void	(*stop)(struct fid_mem_monitor *monitor);
	int	(*subscribe)(struct fid_mem_monitor *monitor,
			const void *addr, size_t len);
	void	(*unsubscribe)(struct fid_mem_monitor *monitor,
			const void *addr, size_t len);
	bool	(*valid)(struct fid_mem_monitor *monitor,
			const void *addr, size_t len);
};
``` 

Among the two ways of change mentioned above, I prefer No. 1 because it is an intuitive incremental change without the need of a refactor of all the function implementations. And we already have one more argument `hmem_info` in `ofi_mem_monitor` API compared to the `fi_ops_mem_monitor` API. So this change does not make the situation too "worse". The only advantage of No.2 may be the benefits of reducing future change in the API if we want to introduce more arguments.

But again, even with No. 1 we still need to change the monitor API. If we cannot change this API then we need to do the mr_entry validation outside `ofi_mr_cache_search`: by using `ofi_mr_cache_find` to query the mr_entry, then do the validation in separate function that does not need to follow the monitor API.

We would like to know what is the feedback from the community before moving forward.

shefty commented:
struct fi_ops_mem_monitor cannot change.  I'm not worried there; we may just need to convert the calls.

Isn't adding ipc_handle to ofi_mr_hmem_info sufficient?  I thought ofi_mr_hmem_info::ipc_handle would be set by subscribe(), similar to hmem_info::xxx_id.  valid() would perform a similar lookup and fail if the stored ipc_handle wasn't the same.

In your example, where did the ipc_handle that you passed into valid() come from?

shijin-aws commented:
@shefty 

>Isn't adding ipc_handle to ofi_mr_hmem_info sufficient? I thought ofi_mr_hmem_info::ipc_handle would be set by subscribe(), similar to hmem_info::xxx_id. valid() would perform a similar lookup and fail if the stored ipc_handle wasn't the same.

Yes, you are right. I forgot to add that in the earlier comment. There should be a `subscribe` call to assign the `ipc_handle` to `entry->hmem_info.ipc_handle`. It will look like this

```
tatic int ipc_handle_monitor_subscribe(struct ofi_mem_monitor *monitor, const void *addr,
					size_t len, void **ipc_handle,
					union ofi_mr_hmem_info *hmem_info)
{
        memcpy((void **)&hmem_info->ipc_handle, ipc_handle, IPC_HANDLE_SIZE);
	return FI_SUCCESS;
}
```

shijin-aws commented:
>In your example, where did the ipc_handle that you passed into valid() come from?

We propose to introduce `ipc_handle` into `ofi_mr_info` as 

```
struct ofi_mr_hmem_ipc_info {
    uint8_t  handle[IPC_HANDLE_SIZE];
    void     *mapped_addr;
};

struct ofi_mr_info {
	struct iovec iov;
	enum fi_hmem_iface iface;
	uint64_t device;
	struct ofi_mr_hmem_ipc_info ipc_info;
};
```

Then we change `ofi_mr_cache_search` API to 

```
int ofi_mr_cache_search(struct ofi_mr_cache *cache, const struct ofi_mr_info *info,
			struct ofi_mr_entry **entry)
```

Caller constructs `info` and assign `ipc_handle` in it before `ofi_mr_cache_search`.

shijin-aws commented:
bot:aws:retest

shijin-aws commented on include/ofi_shm.h:
```diff
@@ -68,7 +69,7 @@ extern "C" {
 #define SMR_FLAG_IPC_SOCK (1 << 2)
 
-#define SMR_CMD_SIZE		128	/* align with 64-byte cache line */
+#define SMR_CMD_SIZE		256	/* align with 64-byte cache line */
```
This change is due to extended `ipc_info` struct, after discussing with @aingerson

shefty replied:
Doesn't this break the SMR protocol, needing a version bump?

shijin-aws replied:
Bumped `SMR_VERSION` to 3.

shefty commented on include/ofi_mr.h:
```diff
@@ -146,9 +153,11 @@ struct ofi_mem_monitor {
 			 union ofi_mr_hmem_info *hmem_info);
 	void (*unsubscribe)(struct ofi_mem_monitor *notifier,
 			    const void *addr, size_t len,
+			    void **ipc_handle,
```
Do we need to modify subscribe/unsubscribe?  Won't these be no-ops for the IPC case, and we always need to call valid() instead?

Can we separate the changes needed to support IPC handles on the exporter side from the importer side?

shijin-aws replied:
`subscribe` is not a no-op for IPC, this is the function

```
static int ipc_handle_monitor_subscribe(struct ofi_mem_monitor *monitor, const void *addr,
					size_t len, void **ipc_handle,
					union ofi_mr_hmem_info *hmem_info)
{
        memcpy((void **)&hmem_info->ipc_handle, ipc_handle, IPC_HANDLE_SIZE);
	return FI_SUCCESS;
}


```

shijin-aws replied:
>Can we separate the changes needed to support IPC handles on the exporter side from the importer side?

Just to make sure I understand, currently the `ofi_import_monitor` follows the same API with `ofi_mem_monitor`, so I need to modify the API for `ofi_import_monitor_valid` and `ofi_import_monitor_subscribe` . This PR does not change the `export_ops` for the exporter side, just ignores the `ipc_handle` when calling the `export_ops` inside `ofi_import_monitor_*`.

Are you suggesting making the changes on import_monitor API as a separate commit, or something else?

shefty commented:
I think I'm still missing the full flow here.  The ipc_handle is used on both sides of communication, correct?  What I'm looking for is a separation between how to improve the flow at the initiator side separate from the target.

Is the application expected to register HMEM buffers and provide a descriptor as input?  (This is what it looks like to me.)  Is there any intent to add a cache at the initiator side to remove the registration requirement?  If that is done, then I think the ipc_handle ends up being opened internally by the cache and given as *output*.  The ipc_handle would be included in the ofi_mr_entry.

The changes here appear to be target side driven.  I can think of 2 reasonable approaches here.  One is to build the ipc cache as wrappers over the ofi_mr_cache calls.  For example:
```
struct ofi_mr_entry * ofi_ipc_cache_find(...)
{
    lock();
    entry = ofi_mr_cache_find(...);
    if (!ipc_handle_valid(entry->ipc_handle))
        update entry->ipc_handle;
    unlock();
    return entry;
}
```
There's basically some additional data that we're wanting to maintain with each ofi_mr_entry, and rather than trying to push the knowledge of that data down into the core MR cache code, instead manage it above.  But we have the cache allocate space for whatever that data is (in this case the ipc_handle) as an optimization.  (A clean layering would have a function like the above return, say, struct ofi_ipc_entry.  But we want this fast, so we share the structure.)

As an alternative, you're wanting to associate each ofi_mr_entry with some additional data.  In this case, the ipc_handle, that the monitor knows how to interpret.  But pushing the copying of the ipc_handle into the monitor subscribe() callback feels off to me.  I guess it's used by the monitor valid() callback, so it's not too bad.  The call path for handling the ipc_handles just get buried in a way that seems more difficult to follow.

shijin-aws commented:
@shefty

>I think I'm still missing the full flow here. The ipc_handle is used on both sides of communication, correct? What I'm looking for is a separation between how to improve the flow at the initiator side separate from the target.

The ipc cache will only be used on the target side. The initiator side who calls `ofi_hmem_get_handle` will not involve cache at all because it's a cheap call. 

>Is the application expected to register HMEM buffers and provide a descriptor as input? (This is what it looks like to me.) Is there any intent to add a cache at the initiator side to remove the registration requirement? If that is done, then I think the ipc_handle ends up being opened internally by the cache and given as output. The ipc_handle would be included in the ofi_mr_entry.

I do not think the memory registration activity is related to the ipc cache here. And there is nothing done in this PR to remove the registration requirement on the initiator side.

>The changes here appear to be target side driven. I can think of 2 reasonable approaches here. One is to build the ipc cache as wrappers over the ofi_mr_cache calls. For example:
struct ofi_mr_entry * ofi_ipc_cache_find(...)
{
    lock();
    entry = ofi_mr_cache_find(...);
    if (!ipc_handle_valid(entry->ipc_handle))
        update entry->ipc_handle;
    unlock();
    return entry;
}
There's basically some additional data that we're wanting to maintain with each ofi_mr_entry, and rather than trying to push the knowledge of that data down into the core MR cache code, instead manage it above. But we have the cache allocate space for whatever that data is (in this case the ipc_handle) as an optimization. (A clean layering would have a function like the above return, say, struct ofi_ipc_entry. But we want this fast, so we share the structure.)

The key point for ipc handle is to invalidate the stale ipc handle when the `ofi_mr_cache_find` returns a mr_entry with the matched iov but a different (stale) ipc handle. In this case the code need to close the stale ipc handle and remove it from the cache, before it opening the new handle and insert the mr entry into the cache.  In the current PR this is done by introducing the ipc monitor, and let `ofi_mr_cache_search` to call `ofi_mr_uncache_entry` when the `valid` op returns negative. And I was concerned that this implementation will cause a deferred `delete_region` because `ofi_mr_uncache_entry` will just add the entry into the dead region list without purging it really. And the entry is actually purged in the subsequent `ofi_mr_cache_flush` call in the next `ofi_mr_cache_search` call.

I think adding a wrapper call like `ofi_ipc_cache_search` to do the workflow above independently is better. It should be something like

```
int ofi_ipc_cache_search(...)
{
entry = ofi_mr_cache_find(..)
if (entry) {
   if (entry->ipc_handle == new_ipc_handle) 
      /* cache hit*/
     return ...
   else 
     /* close stale ipc and remove it from the cache */
    close_handle()
    ofi_mr_cache_delete()
}
open_handle()
cache_insert()
}
```

wzamazon commented:
> I think adding a wrapper call like ofi_ipc_cache_search to do the workflow above independently is better. It should be something like

Note that `ofi_ipc_cache_search` will need to eject least recently used entry from the LRU, when the total mem size and total_mr_count  exceeded the limit.

My concern is that the resulted `ofi_ipc_cache_search` will very similar to `ofi_mr_cache_search`, and we end up maintaining both. For example, if the ejection logic changed, we will need change both.

shefty commented:
@shijin-aws - Yes, that makes sense and is what I was proposing as an option.

@wzamazon - Why would the wrapper need to process the LRU list?  The underlying cache should handle that.

wzamazon commented:
> Why would the wrapper need to process the LRU list? The underlying cache should handle that.

Note that `ofi_mr_cache_search` process the LRU list.

But the proposed function `ofi_ipc_cache_search` does not call `ofi_mr_cache_search`.

It call `ofi_mr_cache_search` to check whether a memory region is in cache, and directly call `cache->add_region` to add a new region.

Because `ofi_ipc_cache_search` insert memory region by itself, the function need to process the LRU list.

shijin-aws commented:
I guess @wzamazon 's point is that the `ofi_mr_cache_search` currently handles the lru flush when `ofi_mr_cache_full`, and also "most" operations involved in the `ofi_ipc_cache_search` proposed above:
- entry find
- entry insert
- entry uncache when it is invalid

. Implementing a separate `ofi_mr_cache_search` as wrapper of `ofi_mr_cache_find` will cause code duplication and harder to maintain in the future.

shefty commented:
Why couldn't ipc_search call cache_search?  If the ipc_handle doesn't match or isn't valid, we just need to update the cached handle, rather than remove the cached entry and insert a new one.

wzamazon commented:
> Why couldn't ipc_search call cache_search? If the ipc_handle doesn't match or isn't valid, we just need to update the cached handle, rather than remove the cached entry and insert a new one.

You are right. Update the entry found would be possible. Thank you!

shijin-aws commented:
>Why couldn't ipc_search call cache_search? If the ipc_handle doesn't match or isn't valid, we just need to update the cached handle, rather than remove the cached entry and insert a new one.

I see, so you are actually proposing this

```
int ofi_ipc_cache_search(...)
{
   ofi_mr_cache_search(...., entry);
   if (!ipc_handle_valid(entry->ipc_handle)) {
        close_handle(entry->ipc_handle)
        update entry->ipc_handle;
        open_handle(entry->ipc_handle;);
}
        
}
```

shijin-aws commented:
@shefty one more question, there is a corner case for CUDA ipc: The sending side gets handle from a  block of memory. Then the sending side frees the memory. The sending side then cudaMalloc again and gets the same base address. However, it cudaMalloc a block that is larger than the one in the cache. The cache will return that memory is not found and the ofi_hmem_open_handle will be called again. However, that will fail with cudaErrorAlreadyMapped. Therefore we need to unmap all overlapping regions and retry the open handle. 

To unmap all overlapping regions and retry the open handle, we need to do mostly what `ofi_mr_cache_notify` does:

```
void ofi_mr_cache_notify(struct ofi_mr_cache *cache, const void *addr, size_t len)
{
	struct ofi_mr_entry *entry;
	struct iovec iov;

	cache->notify_cnt++;
	iov.iov_base = (void *) addr;
	iov.iov_len = len;

	for (entry = ofi_mr_rbt_overlap(&cache->tree, &iov); entry;
	     entry = ofi_mr_rbt_overlap(&cache->tree, &iov))
		util_mr_uncache_entry(cache, entry);
}
```
And then do `ofi_mr_cache_flush` to close the handles in the dead_region_list. Do you think I could just call `ofi_mr_cache_notify` directly? My concern is this call seems hooks with memory monitor which is not available in our case.

shefty commented:
We should keep separate a cache at the initiator/sender versus the target/receiver.  I don't think they'll have the same implementation.

For the target, we could call notify directly.  What are the alternatives?  What indicates that the target needs to flush some range of values from the cache?  We could potentially overlap several different regions.

wzamazon commented:
@shijin-aws 

> The cache will return that memory is not found and the ofi_hmem_open_handle will be called again.
> However, that will fail with cudaErrorAlreadyMapped.

I do not understand why there will be the error of cudaErrorAlreadyMapped.

`ofi_mr_cache_search` will remove any region that is overlapping with the input region.

https://github.com/ofiwg/libfabric/blob/7079d5bad3489330d4741b4b810cecd8ae7a3f37/prov/util/src/util_mr_cache.c#L364

In your example, if the input region is larger than a region in cache. The region in cache should be closed. Right?

shijin-aws commented:
@wzamazon 
>In your example, if the input region is larger than a region in cache. The region in cache should be closed. Right?

But that uncache does not really close the handle, right? It will move the entry into the dead_region_list, and it will be flushed only when `ofi_mr_cache_flush` is called, which is in the next `ofi_mr_cache_search` call.

shijin-aws commented:
Then I think an extra uncache for the overlapped region is unneeded, we need to do `ofi_mr_cache_flush` and then re-open

wzamazon commented:
> we need to do ofi_mr_cache_flush and then re-open the handle when hitting this error.

I think meant that `ofi_ipc_cache_search()` should call `ofi_mr_cache_flush` if `ofi_mr_cache_search` encountered error, then call `ofi_mr_cache_search()` again.

something like:

```
int ofi_ipc_cache_search(...)
{
    int ret = ofi_mr_cache_earch(mr_cache, addr, len);
    while (ret == CUDA_MAP_ALREADY) {
             ofi_mr_cache_flush(mr_cache);
           
             ret = ofi_mr_cache_search(mr_cache, addr, len);
   }

  ....
}
```

shijin-aws commented:
@wzamazon would it be better to handle this flush and retry inside `cache->add_region`?

shijin-aws commented:
Latest revision adds a new commit to introduce shm test suites in `fabtests/pytest/shm`, which will run `fi_rdm_tagged_bw`, `fi_rdm_pingpong` and `fi_rma_bw` between cuda-to-cuda, cuda-to-host, host-to-cuda, host-to-host.

wzamazon commented:
> would it be better to handle this flush and retry inside cache->add_region?

Yes, I think so.

In that case `add->region` will call `ofi_mr_cache_flush`, which I think is fine.

wzamazon commented on prov/util/src/util_mr_cache.c:
```diff
@@ -316,29 +316,23 @@ util_mr_cache_create(struct ofi_mr_cache *cache, const struct ofi_mr_info *info,
 }
 
-int ofi_mr_cache_search(struct ofi_mr_cache *cache, const struct fi_mr_attr *attr,
+int ofi_mr_cache_search(struct ofi_mr_cache *cache, const struct ofi_mr_info *info,
```
Do we still need to change `ofi_mr_cache_search`?

shijin-aws replied:
We still need to pass the `IPC_handle` as part of the `mr_entry->info`, so this change is still needed.

shefty replied:
Why is the ipc_handle needing to be passed down?  What is the core MR cache doing with it?

shijin-aws replied:
ipc_handle needs to be set in `ofi_mr_entry->info`, so the `ofi_mr_cache_search` can create such mr_entry from the info, and insert it to the cache via `util_mr_cache_create`, right?

The `fi_mr_attr` does not have ipc_handle (and we cannot modify this public API). Is there any alternative we can set ipc_handle into the mr_entry without changing the `ofi_mr_cache_search` API?

shijin-aws replied:
In our earlier discussion, isn't this code block require the ipc_handle part of the mr_entry in the cache?

```
if (!ipc_handle_valid(entry->ipc_handle))
        update entry->ipc_handle;
```

shefty replied:
Look at the modified implementation of ofi_mr_cache_search()/ find()/ reg().  They don't touch the ipc_handle or access any data that they already didn't have access to.  The ipc_handle is only accessed by the wrapper call ofi_ipc_cache_search().  I don't think the core mr cache needs to change.  We just need to add the ipc fields to ofi_mr_entry, so the ipc cache has somewhere to store the data.

shijin-aws replied:
But currently `ofi_mr_cache_search` construct the `ofi_mr_info` inside the function, by copying the iov from `fi_mr_attr` in the input? If you do not change the api, the constructed `ofi_mr_info` inside the function will not have ipc_handle, right?

shefty replied:
Yes, if possible.  That isolates IPC handling to the ipc level cache.

wzamazon commented on include/ofi_hmem.h:
```diff
@@ -43,6 +43,18 @@
 
+#define IPC_HANDLE_SIZE		64
+
+struct ipc_info {
```
why there are `ipc_info` and `ofi_mr_hmem_ipc_info`?

shijin-aws replied:
The `ipc_info` is used by shm cmd queue. `ofi_mr_hmem_ipc_info` is the minimum ipc info added to ofi_mr_entry->info.

wzamazon replied:
What is the relationship between them? Can we store `ipc_info` in `ofi_mr_entry` too?

wzamazon replied:
It seems that `ipc_info` was `shm_ipc_info`. The reason it is moved to `ofi_hmem.h` is because `ofi_ipc_cache_search` now takes it as input, and use it to construct a `ofi_mr_info`.

I think it is better to remain `ipc_info` as `shm_ipc_info` and keep it inside shm. shm provider prepare `ofi_mr_info` and call `ofi_ipc_cache_search` with it.

shijin-aws replied:
Move `ipc_info` to `smr_ipc_info` in `ofi_shm.h` in the latest revision.

shijin-aws replied:
continuous-integration/appveyor/pr — AppVeyor build failed because ofi_shm.h cannot be compiled on windows. @wzamazon I think it is because I moved ipc_info to ofi_shm.h and I had to include ofi_shm.h in hmem_ipc_cache.c. Maybe I should move ipc_info (and IPC_HANDLE_SIZE) back to ofi_hmem.h

shefty commented on include/ofi_mr.h:
```diff
@@ -54,10 +54,16 @@
+struct ofi_mr_hmem_ipc_info {
+    uint8_t  handle[IPC_HANDLE_SIZE];
+    void     *mapped_addr;
+};
```
Move fields directly into ofi_mr_info and avoid separate structure.  It's only used within ofi_mr_info anyway

shijin-aws replied:
fixed.

shefty commented on include/ofi_mr.h:
```diff
@@ -338,26 +350,48 @@ static inline bool ofi_mr_cache_full(struct ofi_mr_cache *cache)
+ *			region described in info.
+ * @returns On success, returns 0. On failure, returns a negative error code.
+ */
+int ofi_mr_cache_search(struct ofi_mr_cache *cache, const struct ofi_mr_info *info,
```
Move changes to the existing cache API into its own patch.

shijin-aws replied:
With the latest design, there is no change in the existing cache API. I just need to make `ofi_mr_cache_search` initialize the `info` to `{0}` inside the function. And I made it a separate commit.

shefty commented on prov/util/src/ipc_handle_monitor.c:
```diff
@@ -0,0 +1,77 @@
+	.stop = ipc_handle_monitor_stop,
+	.subscribe = ipc_handle_monitor_subscribe,
+	.unsubscribe = ipc_handle_monitor_unsubscribe,
+	.valid = ipc_handle_monitor_valid,
```
Do we have existing no-op options for the callbacks or can we set them to NULL?  If we don't have existing no-op options, move these to generic no-ops instead.

shijin-aws replied:
We cannot set them to NULL because they are called explicitly in `ofi_mr_cache_search`, `util_mr_cache_create`, and `ofi_monitors_update`.

I did not find general no-op functions for `ofi_mem_monitor` yet. It seems all the other ifaces like cuda/ze/rocr all define the no-ops explicitly.

shefty replied:
We should define generic no-ops for these and update the other monitors to use them.

shijin-aws replied:
I see, will do it then.

shijin-aws replied:
Done. I was not able to update other monitors because they have a branching between no-op and ENOSYS according to `HAVE_CUDA/ZE/ROCR`.

shefty reviewed (commented):
Please separate the changes needed by the core cache, then build the ipc cache over it.  That will help show what exactly is happening.

wzamazon commented on include/ofi_mr.h:
```diff
@@ -181,6 +184,18 @@ void ofi_monitor_unsubscribe(struct ofi_mem_monitor *monitor,
 			     const void *addr, size_t len,
 			     union ofi_mr_hmem_info *hmem_info);
 
+int ofi_monitor_start_no_ops(struct ofi_mem_monitor *monitor);
```
`no_ops` should be `no_op`?

shijin-aws replied:
fixed

wzamazon commented:
> With the latest design, there is no change in the existing cache API. I just need to make ofi_mr_cache_search initialize the info to {0} inside the function. And I made it a separate commit.

I have mixed feelings about the new approach. On the one the hand I like that you did not change `ofi_mr_cache_search`, on the other hand `ofi_ipc_cache_search` end up being very similar to `ofi_mr_cache_search`. I guess I am OK either way.

wzamazon commented on src/hmem_ipc_cache.c:
```diff
@@ -0,0 +1,231 @@
+		FI_INFO(&core_prov, FI_LOG_CORE,
+			"ipc cache opens new ipc handle, addr: %p, len: %lu\n",
+			entry->info.iov.iov_base, entry->info.iov.iov_len);
+		ret = ipc_cache_open_handle(cache, entry);
```
Don't you still need to handle the -FI_EALREADY error by calling `ofi_mr_flush` and retry?

shijin-aws replied:
Currently, we call `ofi_mr_cache_flush` when hitting `-FI_ELAREADY`. If we can do `ofi_mr_cache_flush` every time before opening the handle, we should not need to handle `FI_ELAREADY` explicitly then.

wzamazon replied:
I see. It is handled in add_regions.

shijin-aws replied:
Also, @wckzhang proposed an idea of making the length of the mr_region to be only 1 for ipc cach. By this design, even if the receiver got a new IPC handle with the same base address but larger allocation, the ofi_mr_cache_search will regard it as a cache hit, then `ofi_ipc_cache_search` will invalidate the old IPC handle by closing it and open the new ipc handle. So there should be no `FI_EALREADY ` for sure.

I cannot think of a corner case that can make this idea fail.

shijin-aws replied:
@wzamazon 
> I see. It is handled in add_regions.

No, now we have to make `cache->add_region` a no-op since we keep the `ofi_mr_cache_search` API not changed. The handle open are fully managed inside `ofi_ipc_cache_search`, by calling `ipc_cache_open_handle()`

wckzhang commented on include/ofi_hmem.h:
```diff
@@ -134,6 +134,7 @@ int cuda_dev_unregister(uint64_t handle);
 int cuda_get_handle(void *dev_buf, void **handle);
 int cuda_open_handle(void **handle, uint64_t device, void **ipc_ptr);
 int cuda_close_handle(void *ipc_ptr);
+int cuda_get_base_addr(const void *ptr, void **base, size_t *size);
```
get base addr isn't really what this does. It's more like, get address range

j-xiong replied:
@wckzhang The function name follows the definition of `struct fi_hmem_ops`, which was extended sometime ago from getting base_addr to getting base_addr plus size. Renaming can be considered in the future (I don't see a pressing reason to do so, though), but for this PR the function name is consistent with rest of libfabric.

shijin-aws replied:
Yes, this is an existing API and I just followed it.

wckzhang commented on src/hmem_ipc_cache.c:
```diff
@@ -0,0 +1,231 @@
+
+static void ipc_cache_delete_region(struct ofi_mr_cache *cache, struct ofi_mr_entry *entry)
+{
+
```
Extra line

wckzhang commented on src/hmem_ipc_cache.c:
```diff
@@ -0,0 +1,231 @@
+hit:
+	*mapped_addr = entry->info.ipc_mapped_addr;
+	*mr_entry = entry;
+	ofi_genlock_unlock(&cache->domain->lock);
```
can you add a goto out here and return ret instead of FI_SUCCESS, then add a goto on each exit condition?

wckzhang commented on src/hmem_ipc_cache.c:
```diff
@@ -0,0 +1,231 @@
+#include <ofi_shm.h>
+
+
+static int ipc_cache_open_handle(struct ofi_mr_cache *cache, struct ofi_mr_entry *entry)
```
Can you add doxygen style comments?

wzamazon commented:
bot:aws:retest

shijin-aws commented:
continuous-integration/appveyor/pr — AppVeyor build failed because `ofi_shm.h` cannot be compiled on windows. @wzamazon I think it is because I moved `ipc_info` to `ofi_shm.h` and I had to include `ofi_shm.h` in `hmem_ipc_cache.c`. Maybe I should move `ipc_info` (and IPC_HANDLE_SIZE) back to `ofi_hmem.h`

shijin-aws commented:
@shefty Unfortunately we observed a use case that makes the current design (not changing core mr_cache API and monitor API) not work

Currently, `ofi_ipc_cache_search` has the following workflow:

```
ofi_mr_cache_search(..., ipc_handle, base_addr, len, void **mapped_addr)
{
  // construct fi_mr_attr
  ofi_mr_cache_search(... entry) 
if (!entry->info->ipc_mapped_address) {
    // this is a new entry
    open_handle(ipc_handle)
    // assign the handle to entry
    entry->info->ipc_handle = ipc_handle
} else {
   if (entry->info->ipc_handle == ipc_handle)
      // cache hit
      *mapped_addr = entry->info->mapped_addr
   else {
      // close stale handle
      close_handle(entry->info->ipc_handle)
      // update handle in the entry
      entry->info->ipc_handle = ipc_handle
      // open the new handle
      open_handle(entry->info->ipc_handle)
     } 
  }
}
```

This workflow does not really work in the step 2 of the following scenario:

1. sender malloc a memory region, e.g. 0x0 - 0xF, get ipc handle, sends to receiver, receiver call ofi_mr_cache_search to insert the entry into the cache, and then open the handle and set the ipc_handle in the cached mr_entry.
2. sender free the region and re-allocate a smaller region (0x8-0xF), gets a new ipc handle, sends to receiver. The ofi_mr_cache_search will regard it as a cache hit. But the returned mr_entry has a different IPC handle with the new one, so `ofi_ipc_cache_search` will close the stale handle, update the mr_entry->info->ipc_handle, and then open the new one.   The issue is that the mr_iov of the returned mr_entry also needs an update. And that still requires a uncache of the old mr, and a insert of the new mr_entry (with updated iov) in the cache.

Currently, the entry's uncache (util_mr_uncache_entry) and insert (util_mr_cache_create) are static functions in `ofi_mr_cache_search.c`. We can make them public and call them in `ofi_ipc_cache_search`. But that will make the whole `ofi_ipc_cache_search` a 90% copy of `ofi_mr_cache_search` and we strongly do not recommend this solution.

If we can roll back to the original design: 
1. Change `ofi_mr_cache_search` API to make it able to get `ipc_handle` from its caller, via `ofi_mr_info`.
2. Change `ofi_mem_monitor` API to make `ofi_mr_cache_search` to invalidate the stale IPC handle by calling `ipc_monitor->valid`

All of these corner case can be handled by `ofi_mr_cache_search` itself: When the mr_entry is invalidated by `monitor->valid`, `ofi_mr_cache_search` will uncache it (adding to dead region), then create a new entry and insert it to the rb map.

We would like to get your feedback on this issue, thank you.

shefty commented:
The only change to the current cache that you're proposing is to modify search() and find() to take the ofi_mr_info as input instead of fi_mr_attr, correct?

shijin-aws commented:
@shefty we only need to change `ofi_mr_cache_search` API, and the monitor (internal) API

shefty commented:
It doesn't sound like we have a decent alternative.  I just don't want to slow down the use of the cache for normal memory.

shijin-aws commented:
@shefty I do not know how the change we proposed (ofi_mr_cache_search API, monitor API) can slow down the use of the cache for normal memory. There is no extra code added in `ofi_mr_cache_search`, and the change in monitor API will not introduce extra check for other monitor type.

So you are ok with the change since we do not have a decent alternative?

shefty commented:
@shijin-aws  - Yes, go ahead with the prior set of changes as the base.

shijin-aws commented on prov/efa/src/efa_mr.c:
```diff
@@ -527,7 +531,6 @@ static int efa_mr_reg_impl(struct efa_mr *efa_mr, uint64_t flags, void *attr)
 	assert(efa_mr->mr_fid.key != FI_KEY_NOTAVAIL);
 
 	mr_attr->requested_key = efa_mr->mr_fid.key;
-	mr_attr->device.reserved = efa_mr->peer.device.reserved;
```
It's a typo by accident, this change is actually part of  https://github.com/ofiwg/libfabric/pull/7894. I have fixed it

shijin-aws commented on include/ofi_mr.h:
```diff
@@ -160,7 +162,8 @@ struct ofi_mem_monitor {
 	 */
 	bool (*valid)(struct ofi_mem_monitor *notifier, const void *addr,
-		      size_t len, union ofi_mr_hmem_info *hmem_info);
+		      size_t len, void **ipc_handle,
```
Right, if we can change `valid` API to valid(struct ofi_mr_info *info, struct ofi_mr_entry *entry)`. In that case we do not need to introduce `ipc_handle` to `ofi_mr_hmem_info`. We also do not need to modify the `subscribe` API. The only concern I have is such change makes `valid` quite inconsistent with other APIs in `ofi_mem_monitor`. But I do think it could be better.

shijin-aws replied:
I updated the PR with this valid API change.

shijin-aws commented:
@shefty could you review the PR again? Thx

shijin-aws commented:
@shefty @j-xiong @aingerson could I have another review from the community? Thanks!

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -513,11 +514,14 @@ static int smr_progress_ipc(struct smr_cmd *cmd, enum fi_hmem_iface iface,
+	} else {
+		/* Decrement the use_cnt of the ipc mr_entry so it can be cleaned properly */
+		ofi_mr_cache_delete(ep->ipc_cache, mr_entry);
+	}
```
This flow doesn't look correct.  There shouldn't be some special flow for Ze that's different from the common flow.

shijin-aws replied:
The reason for this flow is that this ipc cache is not enabled for Ze, within this PR (because I cannot test it). So I do not intend to change any existing Ze behavior. Currently Ze opens and closes handle each time in `smr_progress_ipc`. So this `ofi_mr_cache_delete` is only called for non Ze path.

shefty commented on src/hmem_ipc_cache.c:
```diff
@@ -0,0 +1,166 @@
+static int ipc_cache_add_region(struct ofi_mr_cache *cache, struct ofi_mr_entry *entry)
+{
+	int ret;
+	ret = ofi_hmem_open_handle(entry->info.iface, (void **)&entry->info.ipc_handle,
```
Add blank line after variable declarations.

shefty commented on src/hmem_ipc_cache.c:
```diff
@@ -0,0 +1,166 @@
+				"Failed to open hmem handle, addr: %p, len: %lu\n",
+				entry->info.iov.iov_base, entry->info.iov.iov_len);
+		}
+	}
```
This function should be:
```
int ret;

ret = ofi_hmem_open...(...);
if (ret == -FI_EALREADY) {
    ofi_mr_flush...
    ret = ofi_hmem_open...
}
if (ret) {
    FI_WARN(...);
}
return ret;
```
Do not use OFI_UNLIKELY.  Programmers almost always get that wrong, and the errors here can happen.

shefty commented on src/hmem_ipc_cache.c:
```diff
@@ -0,0 +1,166 @@
+	return ret;
+}
+
+static void ipc_cache_delete_region(struct ofi_mr_cache *cache, struct ofi_mr_entry *entry)
```
length

shefty commented on src/hmem_ipc_cache.c:
```diff
@@ -0,0 +1,166 @@
+	};
+	struct util_domain *util_domain;
+	int ret;
+	util_domain = container_of(domain, struct util_domain, domain_fid);
```
blank line to separate declarations

shefty commented on src/hmem_ipc_cache.c:
```diff
@@ -0,0 +1,166 @@
+	struct util_domain *util_domain;
+	int ret;
+	util_domain = container_of(domain, struct util_domain, domain_fid);
+	*cache = (struct ofi_mr_cache *)calloc(1, sizeof(struct ofi_mr_cache));
```
calloc does not need a cast.  Use sizeof(*variable), not sizeof(struct blah)

shefty commented on src/hmem_ipc_cache.c:
```diff
@@ -0,0 +1,166 @@
+	*cache = (struct ofi_mr_cache *)calloc(1, sizeof(struct ofi_mr_cache));
+	if (!*cache)
+		return -FI_ENOMEM;
+	(*cache)->add_region = ipc_cache_add_region;
```
blank line after return

shefty commented on src/hmem_ipc_cache.c:
```diff
@@ -0,0 +1,166 @@
+		        fi_strerror(ret));
+		free(*cache);
+		*cache = NULL;
+		return -ret;
```
Use common exit path for failures where cleanup is needed.

shefty commented on src/hmem_ipc_cache.c:
```diff
@@ -0,0 +1,166 @@
+	if (ret) {
+		FI_WARN(&core_prov, FI_LOG_CORE,
+			"ipc cache init failed: %s\n",
+		        fi_strerror(ret));
```
cache_init should have a warning message, so we shouldn't need one here.

shefty commented on src/hmem_ipc_cache.c:
```diff
@@ -0,0 +1,166 @@
+	struct ofi_mr_info info;
+	struct ofi_mr_entry *entry;
+	int ret;
+	info.iov.iov_base = (void *)ipc_info->base_address;
```
blank line after declarations.

base_address likely needs to convert to u64 if it's not already for use in the shm protocol.  In that case, a cast to (uintptr_t) will be needed.

shefty commented on src/hmem_ipc_cache.c:
```diff
@@ -0,0 +1,166 @@
+	info.iov.iov_base = (void *)ipc_info->base_address;
+	info.iov.iov_len = ipc_info->base_length;
+	info.iface = ipc_info->iface;
+	memcpy((void **)&info.ipc_handle, (void **) &ipc_info->ipc_handle, IPC_HANDLE_SIZE);
```
casts to (void **) shouldn't be needed.  line length

shefty commented on src/hmem_ipc_cache.c:
```diff
@@ -0,0 +1,166 @@
+	if (OFI_UNLIKELY(ret))
+		return ret;
+	*mapped_addr = entry->info.ipc_mapped_addr;
+	*mr_entry = entry;
```
If we return entry, why do we need to copy ipc_mapped_addr out separately?

shijin-aws replied:
That was aimed to have a separate `mapped_addr` output for this function, so it can assign the `base` in the same way with `ze_hmem_open_shared_handle` in https://github.com/ofiwg/libfabric/blob/main/prov/shm/src/smr_progress.c#L493 . I can remove it and use `entry->info.mapped_addr`

shefty reviewed (commented):
See comments and check the line length for all changes.

shijin-aws commented on prov/shm/src/smr_progress.c:
```diff
@@ -489,19 +491,18 @@ static int smr_progress_ipc(struct smr_cmd *cmd, enum fi_hmem_iface iface,
-	ptr = base;
-	if (cmd->msg.data.ipc_info.iface == FI_HMEM_ZE)
-		ptr = (char *) ptr + (uintptr_t) cmd->msg.data.ipc_info.offset;
+	ptr = (char *) base + (uintptr_t) cmd->msg.data.ipc_info.offset;
```
I do not know why the old code did not have offset. But it should have.

shijin-aws commented on src/hmem_cuda.c:
```diff
@@ -103,10 +123,36 @@ static struct cuda_ops cuda_ops = {
+	CUcontext data;
+	return ofi_cuPointerGetAttribute(&data, \
+	CU_POINTER_ATTRIBUTE_CONTEXT, (CUdeviceptr) addr) == CUDA_SUCCESS;
+}
```
I am not sure why, but I cannot see the `""` you mentioned, which line?

shefty replied:
There's a random backslash in the code.

shijin-aws replied:
Thank you. Will remove.

shijin-aws commented on src/hmem_cuda.c:
```diff
@@ -82,15 +87,30 @@ static void *cudart_handle;
+ * We want to make sure we find cuMemFree_v2, not cuMemFree.
+ */
+#define STRINGIFY2(x) #x
+#define STRINGIFY(x)  STRINGIFY2(x)
```
I see, did not realize that gni and psm3 also uses this macro.

shefty commented on src/hmem_cuda.c:
```diff
@@ -103,10 +123,36 @@ static struct cuda_ops cuda_ops = {
+	 * cudaDeviceSynchronize() needs to be called
+	 * to ensure data consistency.
+	 */
+	if (cuda_is_addr_allocated_or_registered(dst) \
```
backslash here too -- apparently that character get's tossed as a comment.

shijin-aws commented on prov/util/src/ze_mem_monitor.c:
```diff
@@ -85,8 +84,7 @@ static void ze_mm_unsubscribe(struct ofi_mem_monitor *monitor,
 static bool ze_mm_valid(struct ofi_mem_monitor *monitor,
-			const void *addr, size_t len,
-			union ofi_mr_hmem_info *hmem_info)
+			const struct ofi_mr_info *info, struct ofi_mr_entry *entry)
```
What is the length limit? 80? Would like to enforce it in my IDE so I will not break it in the future.

shijin-aws commented on prov/util/src/ipc_handle_monitor.c:
```diff
@@ -0,0 +1,77 @@
+			const struct ofi_mr_info *info, struct ofi_mr_entry *entry)
+{
+	return (memcmp((void **)&info->ipc_handle, (void **)&entry->info.ipc_handle,
+		IPC_HANDLE_SIZE) == 0) ? true : false;
```
I am not sure if there is a better solution here. We could only compare the actual size of the IPC handle for different hmem ifaces. But that could require an extra variable in `ipc_info`.

shefty replied:
Can a single instance of an IPC cache realistically support multiple hmem interfaces?  We don't need the size of the ipc handle carried in every instance of ipc_info.  We only need the size per iface.  And there should probably be different monitors per iface.

shijin-aws replied:
> Can a single instance of an IPC cache realistically support multiple hmem interfaces? We don't need the size of the ipc handle carried in every instance of ipc_info. 

I do not think a single instance of `ipc_cache` (`ofi_mr_cache`) should support  multiple hmem ifaces at the same time. The current design expects that only 1 hmem iface is being used in shm transfer.

> We only need the size per iface. And there should probably be different monitors per iface.

Currently the ipc_monitor is assigned to `ofi_mem_monitors[FI_HMEM_CUDA]` for cuda. 
```
struct ofi_mem_monitor *memory_monitors[OFI_HMEM_MAX] = {
		[FI_HMEM_CUDA] = ipc_monitor,
	};
```

I agree we should be able to expand it to more hmem ifaces, but I thought they will just be something like

```
struct ofi_mem_monitor *memory_monitors[OFI_HMEM_MAX] = {
		[FI_HMEM_CUDA] = ipc_monitor,
                [FI_HMEM_ZE] = ipc_monitor,
                [FI_HMEM_ROCR] = ipc_monitor,
                ....
	};
```

The current implementation of `ipc_monitor` in `prov/util/src/ipc_handle_monitor.c` is general and not depending on hmem iface. I think you want to make the compare size in `valid()` different for different hmem ifaces, does that mean we want to change `prov/util/src/ipc_handle_monitor.c` to `prov/util/src/cuda_ipc_handle_monitor.c` and create similar files for other hmem ifaces? Not sure if it's the best practice because it's hard for me to think there will be much difference for the ipc monitor implementation for other hmem ifaces besides the ipc handle size.

shijin-aws commented on prov/shm/src/smr_ep.c:
```diff
@@ -1401,6 +1415,14 @@ int smr_endpoint(struct fid_domain *domain, struct fi_info *info,
 		goto err1;
 
+	if (ofi_hmem_is_initialized(FI_HMEM_CUDA)) {
+		ret = ofi_ipc_cache_open(&ep->ipc_cache, domain);
```
I can first check `(ep->util_ep.caps & FI_HMEM)` so the cache will not be created if user did not request HMEM.

shijin-aws replied:
> If the user did not request HMEM, then no cache is needed.

Consider a client has buffer on host and a server has buffer on device. client does not request HMEM, server request HMEM. If they do ping-pong, client still needs to have the cache setup to process the ipc_handle sent from the sender. 

Is this a legal use case? I do not think sender will check if the receiver has requested HMEM before it decides to use ipc, see https://github.com/ofiwg/libfabric/blob/main/prov/shm/src/smr_msg.c#L195-L199

I think the correct fix is to check `ofi_hmem_is_ipc_enabled(FI_HMEM_CUDA)` here, or maybe even move the cache open into `smr_progress_ipc`?

shijin-aws commented on include/ofi_hmem.h:
```diff
@@ -43,6 +43,17 @@
+	uint64_t	iface;
+	uintptr_t	base_address;
+	uint64_t	base_length;
+	uint8_t		ipc_handle[IPC_HANDLE_SIZE];
```
The `device` and `iface` are copied from the current `smr_ipc_info`: https://github.com/ofiwg/libfabric/blob/main/include/ofi_shm.h#L132-L142. I thought `device` is set as `uint64_t` to be consistent with the `device` union in `fi_mr_attr`, which has a `reserved` as uint64_t.

shijin-aws commented on include/ofi_mr.h:
```diff
@@ -338,7 +338,21 @@ static inline bool ofi_mr_cache_full(struct ofi_mr_cache *cache)
-int ofi_mr_cache_search(struct ofi_mr_cache *cache, const struct fi_mr_attr *attr,
+/**
+ * @brief Given an ofi_mr_info (with an iov range, ipc_info)
+ * If the iov range is already registered and validated by the monitor, assign the corresponding
```
fixed.

shijin-aws commented:
@shefty I think I addressed all of your comments except for two:

1. The size of `iface`, `device` in `ipc_info` (u8 u16 u32)
2. How to intelligently compare ipc_handle for different hmem ifaces.

Will update the PR after discussion.

shijin-aws commented:
@shefty It will be appreciated if you can review the PR again. Thanks a lot.

shefty commented on prov/shm/src/smr.h:
```diff
@@ -280,6 +281,8 @@ struct smr_ep {
 	struct smr_queue	unexp_tagged_queue;
 	struct dlist_entry	sar_list;
+	/* cache for use with hmem ipc */
+	struct ofi_mr_cache	*ipc_cache;
```
This looks like the wrong place for a cache.  A cache per fabric or domain object make more sense, not per endpoint.  Even global might be needed for this.  I think other caches end up being per domain.

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -477,31 +478,42 @@ static int smr_progress_ipc(struct smr_cmd *cmd, enum fi_hmem_iface iface,
+					 cmd->msg.data.ipc_info.iface);
+		if (ret)
+			goto out;
+	}
```
Cache creation shouldn't be in a fast path.  An empty cache is just an initialized rbtree.  This should move to fabric/domain creation.

shefty commented on prov/shm/src/smr_progress.c:
```diff
@@ -513,11 +525,17 @@ static int smr_progress_ipc(struct smr_cmd *cmd, enum fi_hmem_iface iface,
+		/* 
+		 * Decrement the use_cnt of the ipc mr_entry
+		 * so it can be cleaned properly 
+		 */
```
Drop the comment -- "use_cnt" is some internal thing within the cache.  Search should simply be paired with delete.

shefty commented on src/hmem_ipc_cache.c:
```diff
@@ -0,0 +1,164 @@
+{
+	ofi_mr_cache_cleanup(cache);
+	free(cache);
+	cache = NULL;
```
Setting cache=NULL here does nothing.

shefty commented on src/hmem_ipc_cache.c:
```diff
@@ -0,0 +1,164 @@
+	info.iov.iov_base = (void *) (uintptr_t) ipc_info->base_address;
+	info.iov.iov_len = ipc_info->base_length;
+	info.iface = ipc_info->iface;
+	memcpy(&info.ipc_handle, &ipc_info->ipc_handle, IPC_HANDLE_SIZE);
```
Add an ipc_handle_size field to the cache and optimize the memcpy and memcmp routines.  This will ensure that we don't access data beyond the array that's valid and provide iface specific optimizations.

shijin-aws replied:
How about we adding `ipc_handle_size` as part of `ipc_info`, and `ofi_mr_info`? Then we will do
```
memcpy(&info.ipc_handle, &ipc_info->ipc_handle, ipc_info->ipc_handle_size);
```
here,  and 

```
memcmp((void **)&info->ipc_handle,
		(void **)&entry->info.ipc_handle,
		entry->info.ipc_handle_size) == 0;
```

in `valid()`.

shefty commented on include/ofi_hmem.h:
```diff
@@ -43,6 +43,26 @@
+	uint64_t	base_length;
+	uint64_t	device;
+	uint64_t	offset;
+	uint8_t		ipc_handle[IPC_HANDLE_SIZE];
```
IPC_HANDLE_SIZE should be MAX_IPC_HANDLE_SIZE, which is what it represents.

shefty commented on prov/util/src/ipc_handle_monitor.c:
```diff
@@ -0,0 +1,81 @@
+	.subscribe = ipc_handle_monitor_subscribe,
+	.unsubscribe = ipc_handle_monitor_unsubscribe,
+	.valid = ipc_handle_monitor_valid,
+};
```
ipc_handle_monitor works fine for now, but I believe we will want hmem iface specific monitors.  (E.g. cuda_ipc_monitor, ze_ipc_monitor, etc.)  The only function out of these that does anything is the valid() call, which is technically an iface check specific.  The other functions should be declared as generic no-op calls, which are not specific to an ipc monitor.  That is, rename ipc_handle_monitor_start -> ofi_mm_start_noop, etc.  Other implementations that are no-ops can simply reference these no-op calls (which should move into a better location).  For example, you can do ```#define rocr_mm_unsubscribe ofi_mm_unsubscribe_noop``` where the definition isn't needed.

See below.

shijin-aws replied:
OK,  I can start from renaming `ipc_handle_monitor.c` to `cuda_ipc_monitor.c`, implement the general no-ops function, and then create `ze/rocr_ipc_monitor.c` and make the valid() as `ENOSYS` for now.

shijin-aws commented on prov/util/src/util_mr_cache.c:
```diff
@@ -350,24 +344,24 @@ int ofi_mr_cache_search(struct ofi_mr_cache *cache, const struct fi_mr_attr *att
-				   &(*entry)->hmem_info))
+                                  (const void *)(*entry)->info.iov.iov_base,
+                                  (*entry)->info.iov.iov_len,
+                                  &(*entry)->hmem_info))
```
This code change is not available in latest revision. Now `monitor-valid` is only one line: `monitor->valid(monitor, info, *entry))`. Probably this is in an intermediate commit.

shefty replied:
This change shows up in patch 3a5227a.  Tabs have been converted to spaces.  Please revert the change.

shijin-aws replied:
Acked.

shijin-aws commented:
@shefty I address all of your comments. To add ipc handle size info to cache/monitor, I created a new hmem ops `.get_ipc_handle_size`: 

```
struct ofi_hmem_ops {
	...
	int (*get_ipc_handle_size)(size_t *size);
};

```

And I call `ofi_hmem_get_ipc_handle_size(iface, &ipc_handle_size)`

before doing `memcpy` in `ofi_ipc_cache_search`. This makes me no need to add additional `ipc_handle_size` to existing structures. In `cuda_ipc_monitor`, I just use `sizeof(cudaIpcMemHandle_t)` directly when `HAVE_CUDA` is defined.

Could you review it again? Thank you.

shijin-aws commented:
@shefty may I know what's the error in Intel CI?

shefty commented:
Restarted Intel CI.  Previous issue was some connectivity problem trying to reach some server dealing with MPICH testing.

shijin-aws commented:
@shefty Thanks, would you mind reviewing this PR again? Thanks.

shefty commented on include/ofi_shm.h:
```diff
@@ -151,7 +139,7 @@ union smr_cmd_data {
 		uint64_t	sar;
 	};
-	struct smr_ipc_info	ipc_info;
+	struct ipc_info	ipc_info;
```
Please align field ipc_info with the other fields in this struct.

shefty replied:
Based on the view in the browser, all of those fields are aligned (to a specific tab stop).  Do you have tabs set correctly?

shijin-aws replied:
I changed to chrome and find the issue you are saying. Fixed in the latest revision.

shefty commented on src/hmem_cuda.c:
```diff
@@ -106,7 +115,29 @@ static struct cuda_ops cuda_ops = {
+	if (ofi_cuPointerGetAttribute(&data, \
+	CU_POINTER_ATTRIBUTE_CONTEXT, (CUdeviceptr) dst) == CUDA_SUCCESS
+	&& ofi_cuPointerGetAttribute(&data,
+	CU_POINTER_ATTRIBUTE_CONTEXT, (CUdeviceptr) src) == CUDA_SUCCESS)
```
indentation on the if statement is wonky. Also, remove the backslash

shijin-aws replied:
Will fix.

shijin-aws replied:
Fixed.

shefty commented on src/hmem.c:
```diff
@@ -349,3 +355,15 @@ bool ofi_hmem_is_ipc_enabled(enum fi_hmem_iface iface)
+		fi_tostr(&iface, FI_TYPE_HMEM_IFACE),
+		fi_strerror(-ret));
+	return ret;
+}
```
Have this function return size_t and always return the size.  The call should not fail.

shijin-aws replied:
If we make it return `size_t`, how do we handle the case of `HAVE_CUDA=0` ? return 0?

shefty replied:
Yes, I would just return 0.  We should never get far enough to ever call this function in the no-op case.

shijin-aws replied:
Thank you! Updated.

shefty reviewed (commented):
Looks good - only a few minor comments.  Thanks!

shijin-aws commented:
bot:aws:retest
